### ISSUE TITLE
Comma in input file title, m4v, encoder parameters

### ISSUE THREAD

**Issue by MillingtonAllen** [BassThatHertz/video-quality-metrics#22]
It seems to be that it fails if there is a comma in input filename.

Ubuntu 20.10

Computing the VMAF...
[AVFilterGraph @ 0x76ffec0] No such filter: 'FILENAME_AFTER_COMMA)/CRF comparison at preset medium/Raw JSON Data/CRF 30.json:log_fmt'
Error initializing complex filters.
Invalid argument                                                                                                                                                                   
Done!
Traceback (most recent call last):
  File "main.py", line 363, in <module>
    main()
  File "main.py", line 203, in main
    create_table_plot_metrics(json_file_path, args, decimal_places, data_for_current_row, graph_filename,
  File "/media/allen/2637acd8-fd84-4e6e-b830-018fc7f8703c/allen/Desktop/video-quality-metrics-master/save_metrics.py", line 12, in create_table_plot_metrics
    with open(json_file_path, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '(ENTIRE_FILENAME.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 30.json'

A more trivial issue is that handbrake's default of unencrypted m4v doesn't work.  If I rename the files, they work fine as they're based off the mp4 codec.

Also, is there any way possible to pass additional parameters to the encoder?  I'm thinking the encoder tunings like psnr, ssim, animation, stillimage, grain, etc.  Or various color bit depths like 8 bit (default), 10 bit, 12 bit.

Thanks for the project.

**Comment by InB4DevOps** [comment]
Simple: get rid of the comma in the filename. 😄  who puts commas in a filename? 😛 
On a serious note: we could check for commas in the filename and exit the program. Or we need to see why ffmpeg thinks that parts of the filename are part of the filter. (we might need to add quotation marks around the filename)

`A more trivial issue is that handbrake's default of unencrypted m4v doesn't work. If I rename the files, they work fine as they're based off the mp4 codec.`

could you please elaborate? Do .m4v files not work but renaming them to mp4 works?

As for SSIM & PSNR VQM already has parameters for those:
```
  -ssim, --calculate-ssim
                        Calculate SSIM in addition to VMAF.
  -psnr, --calculate-psnr
                        Calculate PSNR in addition to VMAF.
```

We might add support for custom parameters later, what do you think, @BassThatHertz ?

**Comment by MillingtonAllen** [comment]
Yeah, I feel like some of this is trivial in that I could just edit the file names.  Normally, I'd mention it in forums as it doesn't feel important enough for a bug report, but I didn't see the option.  The advantage of leaving everything to a bug report section is at least it's all in the same place if you two decide to work on it in the future.

"Do .m4v files not work but renaming them to mp4 works?"
Yes, if I rename it as .mp4, it works.  This is what happens when I run a file as .m4v

[ipod @ 0x63d7500] Could not find tag for codec hevc in stream #0, codec not currently supported in container
Could not write header for output file #0 (incorrect codec parameters ?): Invalid argument
Error initializing output stream 0:0 -- 

encoded 0 frames
Done!
Traceback (most recent call last):
  File "main.py", line 363, in <module>
    main()
  File "main.py", line 185, in main
    transcoded_bitrate = provider.get_bitrate(transcode_output_path)
  File "/media/allen/2637acd8-fd84-4e6e-b830-018fc7f8703c/allen/Desktop/video-quality-metrics-master/utils.py", line 33, in get_bitrate
    bitrate = probe(video_path)['format']['bit_rate']
KeyError: 'bit_rate'

Wikipedia states that .m4v is similar to mp4 with the primary difference being encryption support.  I'm not sure if there are other nuances.

As for ssim, psnr, the way I read that is that it only calculates the psnr and ssim error rates as opposed to applying the psnr / ssim tuning to the video encodes.  I'm encoding powerpoint presentations, which have mostly static content.  PSNR seems to artifact much less than the default tuning for static content, but blocks more where motion is involved.

My current method is to use video-quality-metrics to find a CRF that looks good in default tune and then do an ffmpeg encode using the psnr tune.  Pretty sure I'm using a higher CRF than needed, but at least it isn't notably artifacting the text.

**Comment by InB4DevOps** [comment]
the m4v issue seems to be related to ffmpeg. Can you do a simple encode 

ffmpeg -ss 00:01:00 -i inputfile.m4v -t 5 outputfile.mp4
?

Maybe it's something with your ffmpeg build?

**Comment by MillingtonAllen** [comment]
I was wondering if it was related to ffmpeg.  ffmpeg doesn't have any issues with that command on the m4v though.  It worked for both that command as well as when I specify libx265.

**Comment by InB4DevOps** [comment]
Regarding additional parameters 
Open ffmpeg_process_factory.py in a text editor, scroll to function get_arguments in class EncodingArguments and add your desired parameters below the crf line
just make sure you have the strings comma-separated and quoted.
eg "--tune", "psnr",

Btw: 

psnr and ssim – these are debugging modes to optimize for good PSNR and SSIM values only. Better metrics don't necessarily mean better quality though.

**Comment by InB4DevOps** [comment]
I just found out that m4v does not support HEVC. Either use x264 or rename your files to mp4. ffmpeg can not write HEVC to m4v container.

**Comment by InB4DevOps** [comment]
ffmpeg -ss 00:01:00 -i inputfile.m4v -t 5 -c:v libx265 outputfile.m4v

This doesn't work, am I right? ;)

**Comment by InB4DevOps** [comment]
For transcoding VQM uses the same file extension as the input file so that's the issue.

**Comment by CrypticSignal** [comment]
> I feel like some of this is trivial in that I could just edit the file names. Normally, I'd mention it in forums as it doesn't feel important enough for a bug report, but I didn't see the option.

Whenever you run into an issue or bug, no matter how small, do not hesitate to open an issue. It's users like yourself who help improve this program by bringing issues to light :)

> I just found out that m4v does not support HEVC. **Either use x264** or rename your files to mp4. 

@MillingtonAllen You shouldn't have to do this as you clearly want to use the x265 encoder. I have work until 10:30pm today, but I can push an update after I finish work (or tomorrow) that will use the MPEG-4 container (.mp4) for all transcodes, if the original video has the m4v extension and the x265 encoder is chosen.

**Comment by CrypticSignal** [comment]
I don't start work until 4pm so I have pushed a couple of changes that will fix both issues. With regards to the issue when the file path has a comma, the filename of the original video will be renamed (all commas are replaced by a hyphen).

@MillingtonAllen If you grab the new update, you should no longer have any issues.

**Comment by InB4DevOps** [comment]
> the filename of the original video will be renamed (all commas are replaced by a hyphen).

This is a bad idea. The input file should only be read not changed in any way. 
I'll revert your commit in a PR and present another solution ;-)

**Comment by MillingtonAllen** [comment]
> ffmpeg -ss 00:01:00 -i inputfile.m4v -t 5 -c:v libx265 outputfile.m4v
> 
> This doesn't work, am I right? ;)

Yes, that appears to fail.  I guess that's one of the nuances of .m4v that I didn't know about.

>This is a bad idea. The input file should only be read not changed in any way.
>I'll revert your commit in a PR and present another solution ;-)

Would a good solution be to symbolic link inside of the (filename) folder you make to alias the file instead?  (Haven't looked into the layout of the program yet).

>all commas are replaced by a hyphen

My first guess would have been an underscore, though that's probably because one of my friends likes using them for everything.

**Comment by MillingtonAllen** [comment]
> ffmpeg -ss 00:01:00 -i inputfile.m4v -t 5 -c:v libx265 outputfile.m4v
> 
> This doesn't work, am I right? ;)

I was talking to my underscore friend about it and he suggested the -f mp4 flag.  I found similar on ffmpeg's website suggesting -f mp4 or -f mov for hevc with a .m4v
https://trac.ffmpeg.org/ticket/7685

**Comment by InB4DevOps** [comment]
> > ffmpeg -ss 00:01:00 -i inputfile.m4v -t 5 -c:v libx265 outputfile.m4v
> > This doesn't work, am I right? ;)
> 
> I was talking to my underscore friend about it and he suggested the -f mp4 flag. I found similar on ffmpeg's website suggesting -f mp4 or -f mov for hevc with a .m4v
> https://trac.ffmpeg.org/ticket/7685

No need for that. Once you tell ffmpeg to save as file.mp4 it will select mp4 as format.

**Comment by InB4DevOps** [comment]
btw: I tried it with a video,file.mp4 (note the comma) and it worked without a problem (this was without @BassThatHertz most recent commit).
Could you please send the exact video path and filename so we can better understand what the problem is?

**Comment by InB4DevOps** [comment]
I also tested it on ubuntu 20.10 now with a filename that contains a comma. What else is wrong with your path? :)

**Comment by MillingtonAllen** [comment]
allen@allen-desktop:~/Desktop/video-quality-metrics-master$ python3 main.py -ovp ~/Videos/Presentation\ 1\ Stuff\,\ More\ -\ My\ Media.mp4 -crf 20 22 24 26 28 30 32 34 36 38 40 42 44 46 --interval 60 --clip-length 1 -e x265                                               

Here's some information about the original video:
Filename: Presentation 1 Stuff, More - My Media.mp4
Bitrate: 763 kbit/s
Framerate: 30/1 (30.0) FPS
-----------------------------------------------------------------------------------------------------------
-----------------------------------------------------------------------------------------------------------
Creating a 1 second clip every 60 seconds from /home/allen/Videos/Presentation 1 Stuff, More - My Media.mp4...
-----------------------------------------------------------------------------------------------------------

... (a bunch of clip creation) ...

Creating clip 73 which starts at 01:13:00...
frame=   30 fps=0.0 q=-1.0 Lsize=     655kB time=00:00:00.96 bitrate=5540.6kbits/s speed=7.26x    

-----------------------------------------------------------------------------------------------------------
Concatenating the clips to create the overview video...
frame= 2190 fps=1325 q=-1.0 Lsize=   81103kB time=00:01:12.96 bitrate=9105.4kbits/s speed=44.1x    
Done!
The clips have been deleted as they are no longer needed.
Overview Video: 1-60 (ClipLength-IntervalSeconds).mp4
-----------------------------------------------------------------------------------------------------------
CRF comparison mode activated.
CRF values 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46 will be compared and the medium preset will be used.
-----------------------------------------------------------------------------------------------------------
Transcoding the video with CRF 20...
x265 [info]: HEVC encoder version 3.4
x265 [info]: build info [Linux][GCC 8.3.0][64 bit] 8bit+10bit+12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [info]: Main profile, Level-4 (Main tier)
x265 [info]: Thread pool created using 12 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 3 / wpp(17 rows)
x265 [info]: Coding QT: max CU size, min CU size : 64 / 8
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : hex / 57 / 2 / 3
x265 [info]: Keyframe min / max / scenecut / bias  : 25 / 250 / 40 / 5.00 
x265 [info]: Lookahead / bframes / badapt        : 20 / 4 / 2
x265 [info]: b-pyramid / weightp / weightb       : 1 / 1 / 0
x265 [info]: References / ref-limit  cu / depth  : 3 / off / on
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 2 / 0.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-20.0 / 0.60
x265 [info]: tools: rd=3 early-skip rskip mode=1 signhide tmvp b-intra
x265 [info]: tools: strong-intra-smoothing lslices=6 deblock sao
frame= 2190 fps= 69 q=28.0 Lsize=    8162kB time=00:01:12.90 bitrate= 917.2kbits/s speed=2.31x    
x265 [info]: frame I:     12, Avg QP:19.74  kb/s: 20961.50
x265 [info]: frame P:    578, Avg QP:24.49  kb/s: 2654.09 
x265 [info]: frame B:   1600, Avg QP:27.72  kb/s: 132.03  
x265 [info]: Weighted P-Frames: Y:0.0% UV:0.0%
x265 [info]: consecutive B-frames: 16.3% 10.0% 12.7% 8.3% 52.7% 

encoded 2190 frames in 31.50s (69.53 fps), 911.80 kb/s, Avg QP:26.82
Done!
Computing the VMAF...
[AVFilterGraph @ 0x68d2280] No such filter: 'More - My Media.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 20.json:log_fmt'
Error initializing complex filters.
Invalid argument
Done!
Traceback (most recent call last):
  File "main.py", line 363, in <module>
    main()
  File "main.py", line 203, in main
    create_table_plot_metrics(json_file_path, args, decimal_places, data_for_current_row, graph_filename,
  File "/media/allen/2637acd8-fd84-4e6e-b830-018fc7f8703c/allen/Desktop/video-quality-metrics-master/save_metrics.py", line 12, in create_table_plot_metrics
    with open(json_file_path, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '(Presentation 1 Stuff, More - My Media.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 20.json'

It seems to work if I edit the comma out.  Perhaps it's comma followed by a space?  I'm not exactly sure.

**Comment by InB4DevOps** [comment]
I think the path is the problem. We don't need your change with the comma @BassThatHertz, we much rather need to find a way to get to the correct path and filename from the abovementioned original video path more reliable.. I guess the many `\` to escape the spaces are a problem for our current way of getting the filename..

**Comment by InB4DevOps** [comment]
Check https://stackoverflow.com/a/8384788

**Comment by CrypticSignal** [comment]
I'm on my work break right now so I'm not home yet. I just wanted to implement a quick fix before starting work so he can at least use his file that has a comma in it. I did realise that it may not be the best solution. With that being said, it's not a good idea to have commas in a filename anyway, so I guess my solution where the file is renamed isn't too bad. But perhaps there's a better solution.

**Comment by CrypticSignal** [comment]
In his last comment, the path he specified isn't surrounded by quotes, and it should be considering there's spaces in it.

**Comment by InB4DevOps** [comment]
> I'm on my work break right now so I'm not home yet. I just wanted to implement a quick fix before starting work so he can at least use his file that has a comma in it. I did realise that it may not be the best solution. With that being said, it's not a good idea to have commas in a filename anyway, so I guess my solution where the file is renamed isn't too bad. But perhaps there's a better solution.

I think it is bad because it makes unneeded and possibly unwanted changes. Although we agree that commas in file names are a bad idea others might not agree and want commas. Let them keep their commas :)

**Comment by InB4DevOps** [comment]
> In his last comment, the path he specified isn't surrounded by quotes, and it should be considering there's spaces in it.

In linux you don't need the quotes. you can escape spaces with `\`
We need to handle the path correctly.

now enjoy your break :)

**Comment by MillingtonAllen** [comment]
> I think the path is the problem. We don't need your change with the comma @BassThatHertz, we much rather need to find a way to get to the correct path and filename from the abovementioned original video path more reliable.. I guess the many `\` to escape the spaces are a problem for our current way of getting the filename..

If I edit the comma out, it seems to work fine.  Perhaps it's the comma followed by a space that's throwing things off.  If I use ' around the filename, it still has trouble with the filename.

**Comment by InB4DevOps** [comment]
[AVFilterGraph @ 0x68d2280] No such filter: 'More - My Media.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 20.json:log_fmt'

since "log_fmt" only appears in function run_libvmaf right after json_file_path, it seems that json_file_path is broken.

**Comment by InB4DevOps** [comment]
do you have a folder (Presentation 1 Stuff, More - My Media.mp4) beside main.py?
if not which folders do you have there?

**Comment by InB4DevOps** [comment]
last comment for today: maybe we shouldn't use the filename as the outputfolder name :)

**Comment by InB4DevOps** [comment]
OR remove the comma from the output_folder. That might be the fix.

**Comment by MillingtonAllen** [comment]
>do you have a folder (Presentation 1 Stuff, More - My Media.mp4) beside main.py?
>if not which folders do you have there?

> [AVFilterGraph @ 0x68d2280] No such filter: 'More - My Media.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 20.json:log_fmt'
> 
> since "log_fmt" only appears in function run_libvmaf right after json_file_path, it seems that json_file_path is broken.

allen@allen-desktop:~/Desktop/video-quality-metrics-master$ python3 main.py -ovp ~/Videos/Presentation\ 1\ Stuff\,\ More\ -\ My\ Media.mp4 -crf 20 22 24 26 28 30 32 34 36 38 40 42 44 46 --interval 60 --clip-length 1 -e x265

produces this before erroring out:

/home/allen/Desktop/video-quality-metrics-master/(Presentation 1 Stuff, More - My Media.mp4)
>1-60 (ClipLength-IntervalSeconds).mp4
>CRF comparison at preset medium
>>Raw JSON Data (empty folder)
>>CRF 20.mp4
>>Table.txt

**Comment by MillingtonAllen** [comment]
Sorry, formatting issue with my last post due to github using the arrow for quotes.  I believe the nesting is as expected.

**Comment by BassThatHertz** [comment]
The following part of the output tells us what the problem is:
`[AVFilterGraph @ 0x68d2280] No such filter: 'More - My Media.mp4)`

Since subprocess.run() takes a list of arguments, and each item in the list is separated by a comma, it seems like the part after the comma is being seen as a filter that we're trying to supply.

Anyway, I can see `Filename: Presentation 1 Stuff, More - My Media.mp4` in the output that @MillingtonAllen pasted. This means that he doesn't have the latest version of this program where commas are replaced by hyphens, because you would see `Filename: Presentation 1 Stuff- More - My Media.mp4` if that were the case. @MillingtonAllen, use the latest version and report back.

Even if there is no issue with the latest version, I don't like this solution of auto-renaming the file by replacing commas with hyphens, I just wanted to implement a quick fix. I think I'll change the code to instead throw an error (and tell the user to rename the file) before exiting the program, when a comma is detected in the path.

**Comment by CrypticSignal** [comment]
@MillingtonAllen 
Just pushed an update that will prevent the name of the output folder from containing any commas. This should eradicate your issue, but let me know if it does not.

**Comment by CrypticSignal** [comment]
I've found another solution which I've just implemented. See commit https://github.com/BassThatHertz/video-quality-metrics/commit/f56ed021a6cd1faf257cd8cd0127d0ad1d55ed9a.

The issue was that if you want to use multiple FFmpeg filters, you separate them with a comma. And, as the comma wasn't being escaped, FFmpeg was looking for a filter named 'More - My Media.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 20.json:log_fmt'. As this filter does not exist, FFmpeg threw an error.

When using an FFmpeg filter (VQM uses the `lavfi` filter), arguments to the filter are specified as `key=value` pairs. In this case, the `value` of the `log_path` key was set to `(Presentation 1 Stuff, More - My Media.mp4)/CRF comparison at preset medium/Raw JSON Data/CRF 20.json`. The comma was not escaped. See https://ffmpeg.org/ffmpeg-filters.html#Notes-on-filtergraph-escaping (I'm not sure if one `\` is needed or two? I didn't fully understand when reading this).

**Comment by MillingtonAllen** [comment]
Both m4v and comma proceeded by space in file name succeeded.

To summarize:

input file title comma issue: solved
m4v file input: solved 
custom encoder parameters: Manual editing of ffmpeg_process_factory.py for now.  Mentioned in this thread:
>InB4DevOps: "We might add support for custom parameters later, what do you think, BassThatHertz?"

I guess I'll close the bug report.  Thanks again!
